File: flyweight/js/MazeGame.js
function MazeGame() {
	this.componentFactory = new ComponentFactory();

	this.createMaze = function(x, y) {
		var m = this.makeMaze(x, y);

		for (var i = 0; i < x; i++) {
			for (var j = 0; j < y; j++) {
				m.addComponent(this.makeWall(), i, j);
			}
		}
		m.addComponent(this.makeRoom(), 1, 2);
		m.addComponent(this.makeRoom(), 2, 2);
		m.addComponent(this.makeRoom(), 3, 2);
		m.addComponent(this.makeRoom(), 1, 1);
		m.addComponent(this.makeRoom(), 3, 1);
		m.addComponent(this.makeRoom(), 1, 3);
		m.addComponent(this.makeRoom(), 3, 3);

		return m;
	}

	this.makeMaze = function(x, y) {
		return new Maze(x, y, (Math.random() * 100));
	}

	this.makeRoom = function() {
		return this.componentFactory.createComponent("coinroom");
	}

	this.makeWall = function() {
		return this.componentFactory.createComponent("wall");
	}
}

function Maze(x, y, s) {
	var length = x;
	var height = y;
	var map = new Array(x);
	var seed = s;

	for (var i = 0; i < x; i++) {
		map[i] = new Array(y);
		for (var j = 0; j < y; j++) {
			map[i][j] = 0;
		}
	}

	this.addComponent = function(component, x, y) {
		if (!(x >= length || y >= height || x < 0 || y < 0)) {
			map[x][y] = component;
		}
	}

	this.printMaze = function() {
		var context = new MazeContext(seed);
		for (var y = 0; y < 5; y++) {
			for (var x = 0; x < 5; x++) {
				process.stdout.write(map[x][y].getSymbol(context));
				context.next();
			}
			process.stdout.write("\n");
		}
	}
}

function ComponentFactory() {
	var flyweights = {};

	this.createComponent = function(componentType) {
		if (!flyweights[componentType]) {
			flyweights[componentType] = new MazeComponent(componentType);
		}

		return flyweights[componentType];
	}
}

function MazeComponent(ct) {
	var componentType = ct;

	this.isSolid = function() {
		if (componentType == "wall") {
			return true;
		} else {
			return false;
		}
	}

	this.getSymbol = function(mazeContext) {
		if (componentType == "wall") {
			return '#';
		} else if (componentType == "room") {
			return '.';
		} else if (componentType == "coinroom") {
			return mazeContext.hasCoin() ? '@' : '.';
		} else {
			return ' ';
		}
	}
}

function MazeContext(s) {
	var seed = s;
	var index = 0;

	this.next = function() {
		index++;
	}

	this.hasCoin = function() {
		var r = new Random(seed*(index+1));
		var coin = false;

		if (r.random() >= 75) {
			return true;
		} else {
			return false;
		}
	}
}

function Random(s) {
	var seed = s;

	this.random = function() {
		var x = Math.sin(seed++) * 10000;
		return (x - Math.floor(x)) * 100;
	}
}

var mg = new MazeGame();
var m = mg.createMaze(5, 5);
m.printMaze();

File: flyweight/java/MazeGame.java
import java.util.Random;

public class MazeGame {

    ComponentFactory cf;

    public MazeGame() {
        cf = new ComponentFactory();
    }

    public Maze createMaze() {
        Maze m = makeMaze(5, 5);

        for (int x = 0; x < 5; x++) {
            for (int y = 0; y < 5; y++) {
                m.addComponent(makeWall(), x, y);
            }
        }

        m.addComponent(makeRoom(), 1, 2);
        m.addComponent(makeRoom(), 2, 2);
        m.addComponent(makeRoom(), 3, 2);
        m.addComponent(makeRoom(), 1, 1);
        m.addComponent(makeRoom(), 3, 1);
        m.addComponent(makeRoom(), 1, 3);
        m.addComponent(makeRoom(), 3, 3);

        return m;
    }

    public Maze makeMaze(int x, int y) {
        return new Maze(x, y);
    }

    public MazeComponent makeRoom() {
        return cf.createComponent(ComponentType.COINROOM);
    }

    public MazeComponent makeWall() {
        return cf.createComponent(ComponentType.WALL);
    }

    public static void main(String[] args) {
        MazeGame mg = new MazeGame();
        Maze m = mg.createMaze();
        m.printMaze();
        m.printMaze();
    }
}

class Maze {
    private MazeComponent[][] map;
    private int length;
    private int height;
    private int seed;

    public Maze(int x, int y) {
        map = new MazeComponent[x][y];
        length = x;
        height = y;

        Random r = new Random();
        seed = r.nextInt();
    }

    public void addComponent(MazeComponent c, int x, int y) {
        if (!(x >= length || y >= height || x < 0 || y < 0)) {
            map[x][y] = c;
        }
    }

    public void printMaze() {
        MazeContext mc = new MazeContext(seed);

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < length; x++) {
                System.out.print(map[x][y].getSymbol(mc));
                mc.next();
            }
            System.out.println("");
        }
    }
}

class ComponentFactory {
    private MazeComponent[] flyweights;

    public ComponentFactory() {
        flyweights = new MazeComponent[ComponentType.values().length];
    }

    public MazeComponent createComponent(ComponentType ct) {
        if (null == flyweights[ct.ordinal()]) {
            flyweights[ct.ordinal()] = new MazeComponent(ct);
        }

        return flyweights[ct.ordinal()];
    }
}

class MazeComponent {

    ComponentType type;
    
    public MazeComponent(ComponentType ct) {
        this.type = ct;
    }

    public boolean isSolid() {
        return type == ComponentType.WALL ? true : false;
    }

    public char getSymbol(MazeContext mc) {
        switch (type) {
            case WALL:
                return '#';
            case ROOM:
                return '.';
            case COINROOM:
                return mc.hasCoin() ? '@' : '.';
            default:
                return ' ';
        }
    }
}

class MazeContext {
    private int seed;
    private int index = 0;

    public MazeContext(int seed) {
        this.seed = seed;
    }

    public void next() {
        index++;
    }

    public boolean hasCoin() {
        Random r = new Random(seed*index);
        if (r.nextInt(100) >= 75) {
            return true;
        } else {
            return false;
        }
    }
}
